broadinstitute/single_cell_portal_core
# Bypass file upload by supplying GS URL or bucket path (SCP-5392)

#### BACKGROUND
Currently, if a user doesn't want to upload a file directly through the SCP UI, they can use the `sync` feature.  This allows them to register files in SCP that have already been pushed to the workspace bucket.  However, there are certain file types and actions that are not supported via `sync`, most notably parsing AnnData files, or author-generated differential expression outputs.  To provide these files, the user is forced to upload them through the UI, which is time-consuming and prone to errors for very large files (as AnnData files tend to be).

#### CHANGES
This update now gives users the ability to specify a path (or GS URL) to a file that already exists in the workspace bucket directly through the upload wizard and bypass the upload process completely.  This can be used for _any_ file form, including the "AnnData UX".  The form will accept either a relative path to the file from the bucket root, or a fully-qualified GS URL.  This value is then used to access the remote file and validate that it exists before saving the new `StudyFile` record.  Additionally, some helper features have been added to browse the study bucket (for copying paths or URLs), and to link to our [documentation](https://singlecell.zendesk.com/hc/en-us/articles/360061006011) on how to upload large files directly to buckets. 

Additionally, client-side file validation (CSFV) is still enabled for uploads using this feature.  This was achieved by adapting the same paradigm used in `sync` to read the remote file.  If the file fails validation, the normal error message UX shows, and the save button is disabled.  Example:

![Screenshot 2024-01-10 at 12 20 32 PM](https://github.com/broadinstitute/single_cell_portal_core/assets/729968/6c369f37-e90d-4c60-bdf0-007307148e0f)

In the case where the file cannot be found - either it does not exist, or if the user does not have permission to view the file client-side (will be true in private studies for non-Terra users) - instead of showing an error, a warning will be issued noting that the file cannot be accessed, and will be validated completely on save.  This allows this feature to be used by the large cross section of private study, non-Terra users, and mimics the same functionality of very large uploads in CSFV, where validation is either too expensive, or impossible.  Example message:

![Screenshot 2024-01-10 at 12 20 12 PM](https://github.com/broadinstitute/single_cell_portal_core/assets/729968/20d04120-ee76-4558-ad38-b5215ade3dda)

The user can still save the file, and normal server-side validations will enforce, which include validating that the file exists in the bucket.  If the file still cannot be found, the save will fail and the following message displayed to the user:

![Screenshot 2024-01-10 at 2 51 47 PM](https://github.com/broadinstitute/single_cell_portal_core/assets/729968/023e5ec9-3a0d-49a3-8517-828cdf74a467)

Demo of the entire UX, including documentation/bucket links and CSFV:

https://github.com/broadinstitute/single_cell_portal_core/assets/729968/f4d40ed8-846c-42a4-aa8a-7c426827c2d2

Lastly, if a user deletes a file through the UI that used this feature, the file will not be removed from the bucket.  This mirrors the same functionality in `sync`, and prevents removing potentially large files that failed ingest for non-deterministic reasons.  The following message is displayed instead:

![new_delete_confirm](https://github.com/broadinstitute/single_cell_portal_core/assets/729968/48a4932b-f18a-49a3-877b-1d62cf24c3e5)

#### MANUAL TESTING
1. Boot all services and sign in
2. Go to the upload wizard for any study (use `Classic` mode for new studies)
3. Go to the Clustering tab, and click the `Use bucket path` button next to the upload control
4. Click the `Browse bucket` link, then add the files `test/test_data/cluster_example.txt` and `test/test_data/cluster_bad.txt` to the bucket
5. Back in the upload wizard, supply the value `cluster_bad.txt` in the field for the bucket path
6. Confirm you see the same error message as shown above, and that the file save is disabled
7. Next, provide the name `no_file.txt`, and confirm that the warning `Unable to access the requested file` is displayed
8. Give the cluster a name and then click `Save`, confirming that the save fails with `no file found at no_file.txt` error
9. Now specify the fully qualified GS URL for `cluster_example.txt` like shown in the demo and then click `Save`, confirming that the action succeeds, and there is now a link to download the file
10. Once the ingest completes, click the `Delete` button and confirm you see the same dialog as above, then delete the file
11. Back in the bucket window, refresh the page and confirm the file is still present
12. (Optional) Try the same with any other file type and confirm you are able to validate & save files once they have been added to the bucket
